geo/src/algorithm/concave_hull.rs

```diff
@@ -0,0 +1,433 @@
+use crate::algorithm::euclidean_distance::EuclideanDistance;
+use crate::algorithm::euclidean_length::EuclideanLength;
+use crate::prelude::Centroid;
+use crate::utils::partial_min;
+use crate::{Line, LineString, MultiLineString, MultiPoint, MultiPolygon, Point, Polygon};
+use num_traits::Float;
+use rstar::{RTree, RTreeNum};
+use std::collections::VecDeque;
+use geo_types::{CoordinateType, Coordinate};
+use crate::algorithm::convex_hull::qhull;
+use crate::algorithm::kernels::HasKernel;
+
+/// Returns a polygon which covers a geometry. Unlike convex hulls, which also cover
+/// their geometry, a concave hull does so while trying to further minimize its area by
+/// constructing edges such that the exterior of the polygon incorporates points that would
+/// be interior points in a convex hull.
+///
+/// This implementation is inspired by https://github.com/mapbox/concaveman
+/// and also uses ideas from the following paper:
+/// www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf
+///
+/// # Example
+/// ```
+/// use geo::{line_string, polygon};
+/// use geo::algorithm::concave_hull::ConcaveHull;
+///
+/// // a square shape
+/// let poly = polygon![
+///     (x: 0.0, y: 0.0),
+///     (x: 4.0, y: 0.0),
+///     (x: 4.0, y: 4.0),
+///     (x: 0.0, y: 4.0),
+/// ];
+///
+/// // The correct concave hull coordinates
+/// let correct_hull = line_string![
+///     (x: 4.0, y: 0.0),
+///     (x: 4.0, y: 4.0),
+///     (x: 0.0, y: 4.0),
+///     (x: 0.0, y: 0.0),
+///     (x: 4.0, y: 0.0),
+/// ];
+///
+/// let res = poly.concave_hull(2.0);
+/// assert_eq!(res.exterior(), &correct_hull);
+/// ```
+pub trait ConcaveHull {
+    type Scalar: CoordinateType;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar>;
+}
+
+impl<T> ConcaveHull for Polygon<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        let mut points: Vec<_> = self.exterior().points_iter().map(|point|{
+           point.0
+        }).collect();
+        Polygon::new(concave_hull(&mut points, concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for MultiPolygon<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        let mut aggregated: Vec<Coordinate<Self::Scalar>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.iter().map(|c| *c))
+            .collect();
+        Polygon::new(concave_hull(&mut aggregated, concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for LineString<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        Polygon::new(concave_hull(&mut self.clone().0, concavity), vec![])
```

Comment:
```suggestion
        Polygon::new(concave_hull(&mut self.0.clone(), concavity), vec![])
```

Pretty nitpicky - but since we only need the coords, we can clone just those, rather than cloning the entire Geometry.